text/0000-machine-input.md

```diff
@@ -0,0 +1,183 @@
+- [GitHub Actions inputs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs)
+
+## Detailed design
+
+**Overall API**
+
+This RFC proposes to extend `{ type: 'xstate.init' }`, which is always the first event that a machine receives, with the optional `data?: ...` property:
+
+```ts
+const greetingMachine = createMachine({
+  schema: {
+    // For type-safety; can be used with JSON Schema
+    input: {} as { name: string }
+  },
+  entry: (_, event) => {
+    // event.type is 'xstate.init'
+    // event.data is { name: ... }
+    console.log(`Hello, ${event.data.name}!`);
+  }
+});
+
+// Non-interpreted environments
+const greetingMachineWithInput = greetingMachine
+  .withInput({ name: 'David' });
+
+greetingMachineWithInput.initialState.event.data;
+// { name: 'David' }
+
+// Interpreted environments
+interpret(greetingMachineWithInput).start();
+// logs 'Hello, David!'
+```
+
+### Technical Background
+
+The reason this was not implemented before is because SCXML does not define a separation between input data (params) and context data (datamodel). To provide "input data" to a statechart in SCXML, you provide [`<param>` values](https://www.w3.org/TR/scxml/#param) as [children inside of `<invoke>`](https://www.w3.org/TR/scxml/#N10FF5), which will overwrite the values in the `<datamodel>`.
+
+The main problem with this is that there is no difference between datamodel values that should only be present at the initialization of a machine and those that can change (via `<assign>`) via transitions in the machine.
+
+### Implementation
+
+**Input (init) event**
+
+```ts
+interface InitEvent<TInputData> {
+  type: 'xstate.init';
+  data: TInputData
+}
+```
+
+**`.withInput(...)` method**
+
+`machine.withInput(input)`
+
+Arguments:
+- `input`: the input data to include in the `data` property of the `InitEvent`
+
+Returns:
+- `machineWithInput`: a cloned `StateMachine` instance with the input data specified
+
+## How we teach this
+
+- As a better alternative to `machine.withContext({...})`, which as been confusing for many
+- As a new concept?
+
+TODO
+
+## Drawbacks
+
+- The only trade-off with this would be introducing a "new" concept, although this is _not_ a completely new concept to XState (events with data); it's simply extending the initial event with `data`. This is also not a new concept for workflows in general, but rather a missing feature in XState.
+
+## Alternatives
+
+- This was originally proposed in a different form in this input PR: https://github.com/statelyai/xstate/pull/2779 . The design and goals of this PR were different, larger in scope, and for a different purpose than this proposal, which only seeks to enable providing input data _once_ - at the start of a machine.
+
+## Unresolved questions
+
+- Should `data` be an optional property of `InitEvent`?
+- Should we enforce at runtime that input data was passed in? If so, how?
```

Comment:
we probably can't easily do this, but we should strive to ensure this at the type-level